Objects/longobject.c

```diff
@@ -1199,17 +1219,51 @@ PyLong_AsNativeBytes(PyObject* vv, void* buffer, Py_ssize_t n, int endianness)
             _PyLong_AsByteArray(v, buffer, (size_t)n, little_endian, 1, 0);
         }
 
-        // More efficient calculation for number of bytes required?
+        /* Calculates the number of bits required for the *absolute* value
+         * of v. This does not take sign into account, only magnitude. */
         size_t nb = _PyLong_NumBits((PyObject *)v);
         /* Normally this would be((nb - 1) / 8) + 1 to avoid rounding up
          * multiples of 8 to the next byte, but we add an implied bit for
          * the sign and it cancels out. */
-        size_t n_needed = (nb / 8) + 1;
-        res = (Py_ssize_t)n_needed;
-        if ((size_t)res != n_needed) {
-            PyErr_SetString(PyExc_OverflowError,
-                "value too large to convert");
-            res = -1;
+        res = (Py_ssize_t)(nb / 8) + 1;
+
+        /* Two edge cases exist that are best handled after extracting the
+         * bits. These may result in us reporting overflow when the value
+         * actually fits.
+         */
+        if (n > 0 && res == n + 1 && nb == n * 8) {
```

Comment:
With `res = (nb / 8) + 1` and `res == n + 1`, the last comparison is just if the bits fill up whole bytes; perhaps that's the best way to avoid the warning:
```suggestion
        if (n > 0 && res == n + 1 && nb % 8 == 0) {
```